src/basic/rlimit-util.c

```diff
@@ -55,6 +61,39 @@ int setrlimit_closest(int resource, const struct rlimit *rlim) {
         return 0;
 }
 
+int setpriority_closest(int nice) {
+        int current, limit;
+        struct rlimit highest;
+
+        // Try to set requested nice level
+        if (setpriority(PRIO_PROCESS, 0, nice) == 0)
+                return 0;
+
+        // Permission failed
+        if (!IN_SET(errno, EPERM, EACCES))
+                return -errno;
+
+        errno = 0;
+        current = getpriority(PRIO_PROCESS, 0);
+        if (errno)
+                return -errno;
+
+        if (getrlimit(RLIMIT_NICE, &highest) < 0)
+                return -errno;
+
+        limit = rlimit_to_nice(highest.rlim_cur);
+
+        // We are already nicer than limit allows us
+        if (current < limit)
+                return 0;
+
+        // Push to the allowed limit
+        if (setpriority(PRIO_PROCESS, 0, limit) < 0)
+                return -errno;
+
```

Comment:
i think adding a log_debug() line here that says "couldn't set nice level to what was asked, picked next best instead..." or so might make sense here...